docs/model_developers/modules/handler.rst

```diff
@@ -0,0 +1,175 @@
+mutually exclusive. If no handler is available, the resource will be marked unavailable.
+
+:class:`~inmanta.agent.handler.ResourceHandler` is the handler base class.
+:class:`~inmanta.agent.handler.CRUDHandler` provides a more recent base class that is better suited
+for resources that are manipulated with Create, Delete or Update operations. This operations often
+match managed APIs very well. The CRUDHandler is recommended for new handlers unless the resource
+has special resource states that do not match CRUD operations.
+
+Each handler basically needs to support two things: reading the current state and changing the state
+of the resource to the desired state in the orchestration model. Reading the state is used for dry
+runs and reporting. The CRUDHandler handler also uses the result to determine whether create, delete
+or update needs to be invoked.
+
+The context (See :class:`~inmanta.agent.handler.HandlerContext`) passed to most methods is used to
+report results, changes and logs to the handler and the server.
+
+Built-in Handler utilities
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The :term:`Inmanta Agent<agent>`, responsible for executing handlers has built-in utilities to help
+handler development. This section describes the most important ones.
+
+Logging
+"""""""
+
+The agent has a built-in logging facility, similar to the standard python logger. All logs written
+to this logger will be sent to the server and are available via the dashboard and the API. 
+Additionally, the logs go into the agent's logfile and into the resource-action log on the server. 
+
+To use this logger, use one of the methods: :py:meth:`ctx.debug<inmanta.agent.handler.HandlerContext.debug>`,
+:py:meth:`ctx.info<inmanta.agent.handler.HandlerContext.info>`,
+:py:meth:`ctx.warning<inmanta.agent.handler.HandlerContext.warning>`,
+:py:meth:`ctx.error<inmanta.agent.handler.HandlerContext.error>`,
+:py:meth:`ctx.critical<inmanta.agent.handler.HandlerContext.critical>` or
+:py:meth:`ctx.exception<inmanta.agent.handler.HandlerContext.exception>`.
+
+This logger supports kwargs. The kwargs have to be json serializable. They will be available via the API in their json structured form.
+
+For example:
+
+.. code-block:: python
+
+    def create_resource(self, ctx: HandlerContext, resource: ELB) -> None:
+        # ...
+        ctx.debug("Creating loadbalancer with security group %(sg)s", sg=sg_id)
+
+
+Caching
+"""""""
+
+The agent maintains a cache, that is kept over handler invocations. It can, for example, be used to
+cache connection, so that multiple resources on the same device can share a connection.
+
+The cache can be invalidated either based on a timeout or on version. A timeout based cache is kept
+for a specific time. A version based cache is used for all resource in a specific version.
+The cache will be dropped when the deployment for this version is ready.
+
+The cache can be used through the :py:func:`@cache<inmanta.agent.handler.cache>` decorator. Any
```

Comment:
Here the @-notation is used to refer to a decorator while in all other places the ()-notation is used. Personally, I like the @-notation more. We should use one notation consistently to prevent confusion.

Reply:
thx, added